### PR TITLE
Faster bilinear resize for x86

### PR DESCRIPTION
**Pull request by MattConley**
- Resubmission of PR 21681
- Added F16C check around _mm_cvtph_ps function
- Integrated changes with half-pixel centers

### CODE REVIEW

**Comment by alextp** [comment]
@penpornk I think you're a better reviewer for this

File: tensorflow/core/kernels/crop_and_resize_op.cc [hunk 1]
```diff
@@ -228,61 +230,58 @@ struct CropAndResize<CPUDevice, T> {
...
-              for (int d = 0; d < depth; ++d) {
-                crops(b, y, x, d) = extrapolation_value;
-              }
-            }
-            continue;
+        if (method_name == "bilinear") {
+          std::vector<CachedInterpolation> xs;
+          std::vector<CachedInterpolation> ys;
+          int min_ix, max_ix, min_iy, max_iy;
+          // TODO: Should the half_pixel_centers be added to this function?
```
Comment (MattConley) [inline tensorflow/core/kernels/crop_and_resize_op.cc]: Just wanted to highlight whether half-pixel center logic would be useful here

**Comment by penpornk** [comment]
@MattConley Apologies for my delay! I was on vacation last week and I'm still catching up with other stuff. The PR looks awesome and I hope to get to it in a few days. Thank you for contributing!

**Comment by MattConley** [comment]
@penpornk No problem, thanks for the response!

**Comment by nluehr** [comment]
@penpornk any update on this?

**Comment by tfboyd** [comment]
@penpornk   Are you back and can take a look?

**Comment by penpornk** [comment]
@MattConley @nluehr @tfboyd Apologies again! I'm also on vacation this week. I'll review this by the coming Monday (for real this time).

**Review by penpornk — CHANGES_REQUESTED** [review]
So sorry again for the delay! I've gone through about 1/5 of the PR so far but got enough comments to post for the first round. Most are styling. Apologies if some of the questions I asked is made clear later in the part that I haven't reached.

Could you please post the benchmark numbers just like https://github.com/tensorflow/tensorflow/pull/21681? Thank you!

Also, I think the packing code can be more compact by using `#define` and C++ templating. For example, instead of naming a new function for a different number of bytes, 
```
pack_1byte() { ... }
pack_2byte() { ... }

template <uint8>
pack() { pack_1byte(); }

template <int16>
pack() { pack_2byte(); }

// Call: pack<uint8>();
```
we use the number of bytes as a template parameter
```
template <unsigned int b>
pack() { // Do something based on b. }

// Call: pack<sizeof<uint8>>();
```
Will post more later.

**Comment by MattConley** [comment]
Thanks for the review, I'm making progress with the changes and will push the fixes soon.

File: tensorflow/core/kernels/crop_and_resize_op.cc [hunk 2]
```diff
@@ -228,61 +230,58 @@ struct CropAndResize<CPUDevice, T> {
...
+          compute_interpolation_weights(LegacyScaler(), crop_width, image_width,
+                                        x1, x2, &min_ix, &max_ix, &xs);
+          compute_interpolation_weights(LegacyScaler(), crop_height,
+                                        image_height, y1, y2, &min_iy, &max_iy,
+                                        &ys);
+
+          // multiply by depth to avoid multiplication in resize_single_image.
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_and_resize_op.cc]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 3]
```diff
@@ -0,0 +1,5619 @@
...
+#include <cstdlib>
+#include <list>
+#include <string>
+
+namespace tensorflow {
+namespace {
+
+// Compute the interpolation indices only once.
+struct CachedInterpolation {
+  int lower;  // Lower source index used in the interpolation
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: We assume that there are no images with more than 2 billion pixels per dimension, which allows us to reduce cache pressure and improve performance

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 4]
```diff
@@ -0,0 +1,5619 @@
...
+bool compute_single_interpolation_weight(const Scaler scaler, const int in_size,
+                                         const float out2in_scale,
+                                         const float out2in_start,
+                                         const bool clip, const int i,
+                                         int* lower, int* upper, float* lerp) {
+  // const float in = i * out2in_scale + out2in_start;
+  const float in = scaler(i, out2in_scale) + out2in_start;
+  *lower = (int)floor(in);
+  *upper = (int)ceil(in);
+  *lerp = (float)(in - (float)*lower);
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 5]
```diff
@@ -0,0 +1,5619 @@
...
+    else if (*upper >= in_size)
+      *upper = in_size - 1;
+    return true;
+  } else {
+    return (*lower >= 0 && *upper < in_size) ? true : false;
+  }
+}
+/**
+ * Compute interpolation values for output indexes in range
+ * [out_start,out_start+out_size-1].
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Should be done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 6]
```diff
@@ -0,0 +1,5619 @@
...
+  interpolation[out_size].lower = 0;
+  interpolation[out_size].upper = 0;
+  const bool clip = true;
+  if (!compute_interpolation_weights(scaler, 0, out_size - 1, in_size,
+                                     out2in_scale, 0.0f, clip, interpolation)) {
+    // Should never happen, check for it anyway
+    printf(
+        "Warning! Interpolation values have lower,upper indexes outside of "
+        "range [0,in_size-1]\n");
+  }
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 7]
```diff
@@ -0,0 +1,5619 @@
...
+bool compute_interpolation_weights(
+    const Scaler scaler, const int out_size, const int in_size,
+    const float x1,  // lower bounding box, crop region starts at in_size*x1
+    const float x2,  // upper bounding box, crop region ends at in_size*x2
+    int* min_i, int* max_i, std::vector<CachedInterpolation>* interpolation) {
+  float out2in_start = out_size > 1
+                           ? (float)(in_size - 1) * (float)x1
+                           : (float)(in_size - 1) * (float)(x1 + x2) / 2.0f;
+  float out2in_scale =
+      out_size > 1
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Applied clang-format before pushing the latest changes, and it looked good, thanks

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 8]
```diff
@@ -0,0 +1,5619 @@
...
+          ? (float)(x2 - x1) * (float)(in_size - 1) / (float)(out_size - 1)
+          : 0.0f;
+  if (compute_minmax_indexes(scaler, out_size, in_size, out2in_scale,
+                             out2in_start, min_i, max_i)) {
+    interpolation->resize(*max_i - *min_i + 1);
+    bool all_inputs_ok = compute_interpolation_weights(
+        scaler, *min_i, *max_i, in_size, out2in_scale, out2in_start, false,
+        interpolation->data());
+    if (!all_inputs_ok) {
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 9]
```diff
@@ -0,0 +1,5619 @@
...
+}
+
+/**
+ * Cast float v to type U with range clamping.
+ *
+ * If v<min_val, return value is clamped to u_min_val. similarly if v>max_val,
+ * return value is clamped to u_max_val.
+ */
+template <typename U>
+U cast_to(float v, float min_val, float max_val, U u_min_val, U u_max_val);
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Renamed to clamp_cast, if that works?

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 10]
```diff
@@ -0,0 +1,5619 @@
...
+  return top + (bottom - top) * y_lerp;
+}
+
+/**
+ * Computes the bilinear interpolation from the appropriate 4 float points
+ * and the linear interpolation weights.
+ * Accepts input tensors of type T and produces output tensors of type U.
+ * Optionally flips horizontal and/or vertical axis.
+ */
+template <typename T, typename U>
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: When trying to build with the macro on the function definition, I got this error: 

> ./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
 #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
                               ^
./tensorflow/core/kernels/crop_resize_bilinear_core.h:228:61: note: in expansion of macro 'TF_ATTRIBUTE_NOINLINE'
                               const bool flip_y, U* output) TF_ATTRIBUTE_NOINLINE {

> 

Is there a better way to apply that attribute?  If not, I've left it as is (with the extra declaration) for now.

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 11]
```diff
@@ -0,0 +1,5619 @@
...
+            cast_to<U>(result1, min_val, max_val, u_min_val, u_max_val);
+        output_y_ptr[x * 4 + 2] =
+            cast_to<U>(result2, min_val, max_val, u_min_val, u_max_val);
+        output_y_ptr[x * 4 + 3] =
+            cast_to<U>(result3, min_val, max_val, u_min_val, u_max_val);
+      }
+      output_y_ptr =
+          flip_y ? output_y_ptr - out_row_size : output_y_ptr + out_row_size;
+    }
+  } else {
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: There is a performance benefit to having the explicit cases, though we may be able to reduce duplicated code by using a template with the number of channels as an argument, if that's something we need to do.  Looping in the original author @thorjohnsen if there are further questions.

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 12]
```diff
@@ -0,0 +1,5619 @@
...
+  // convert 8 packed words of type T to fp32.
+  // T must be one of uint8, int8, uint16, int16, int32, Eigen::half, bfloat16
+  // or float.
+  __m256 to_fp32(__m256i raw);
+#else
+  // convert 4 packed words of type T to fp32.
+  // T must be one of uint8, int8, uint16, int16, int32, Eigen::half, bfloat16
+  // or float.
+  __m128 to_fp32(__m128i raw);
+#endif
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 13]
```diff
@@ -0,0 +1,5619 @@
...
+//
+// The remaining code implements explicitly vectorized versions of a bilinear
+// image resizer.
+// Images with 1, 2, 3 or 4 channels are supported.
+// The image resizer reads samples of type T and writes samples of type U.
+// T and U can be any of the following: uint8, int8, uint16, int16, int32,
+// Eigen::half, bfloat16 and float.
+// There are separate codes for SSE4.1 and AVX2. Enabling AVX2 also enables
+// FP16C instruction set,
+// which contains instructions that convert between Eigen::half and float. The
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 14]
```diff
@@ -0,0 +1,5619 @@
...
+  const float bottom = bottom_left + (bottom_right - bottom_left) * x_lerp;
+  return top + (bottom - top) * y_lerp;
+}
+
+/**
+ * Computes the bilinear interpolation from the appropriate 4 float points
+ * and the linear interpolation weights.
+ * Accepts input tensors of type T and produces output tensors of type U.
+ * Optionally flips horizontal and/or vertical axis.
+ */
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Added a comment to that extent.
My understanding is that this fallback code actually handles two cases, so @thorjohnsen unified the code such that the one function handles both.

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 15]
```diff
@@ -0,0 +1,5619 @@
...
+    if (!compute_single_interpolation_weight(
+            scaler, in_size, out2in_scale, out2in_start, clip, i + min_i,
+            &interpolation[i].lower, &interpolation[i].upper,
+            &interpolation[i].lerp)) {
+      rval = false;
+    }
+  }
+  return rval;
+}
+/**
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/crop_resize_bilinear_core.h]: Done

File: tensorflow/core/kernels/resize_bilinear_op.cc [hunk 16]
```diff
@@ -221,14 +83,23 @@ struct ResizeBilinear<CPUDevice, T> {
...
     if (out_height == in_height && out_width == in_width) {
       output = images.template cast<float>();
       return;
     }
 
-    std::vector<CachedInterpolation> ys(out_height + 1);
-    std::vector<CachedInterpolation> xs(out_width + 1);
+    // Compute the cached interpolation weights on the x and y dimensions.
+    std::vector<CachedInterpolation> ys;
+    ys.resize(out_height + 1);
```
[the remark this answers is not in the document]
Comment (MattConley) [reply tensorflow/core/kernels/resize_bilinear_op.cc]: I'm not sure about this; @thorjohnsen is this change for performance?

**Comment by MattConley** [comment]
@penpornk Thanks again for your review, I just pushed up some changes addressing your comments.

**Comment by rmlarsen** [comment]
This code is pretty unreadable. @martinwicke  do you have any guidance? I don't think we can accept this PR as is. @MattConley  Could you reuse parts of Eigen instead of adding your own library of low-level operations using intrinsics directly?

File: tensorflow/core/kernels/crop_and_resize_op.cc (hunk 1 above)
Comment (martinwicke) [reply tensorflow/core/kernels/crop_and_resize_op.cc]: Whatever we do here, it has to be consistent with the behavior in resize_v2.

@johnpjf FYI

**Comment by martinwicke** [comment]
Hmm... I don't see where intrinsics are used directly here? If you cannot read the code that's a bad sign though.

File: tensorflow/core/kernels/crop_resize_bilinear_core.h [hunk 17]
```diff
@@ -0,0 +1,5587 @@
...
+  __m256i extract_right_4ch(const __m256i left);
+#else
+  __m128i extract_right_1ch(const __m128i left);
+  __m128i extract_right_2ch(const __m128i left);
+  __m128i extract_right_3ch(const __m128i left);
+  __m128i extract_right_4ch(const __m128i left);
+#endif  // __AVX2__
+
+#ifdef __AVX2__
+  // Load top left and bottom left interpolation inputs into output argument
```
Comment (penpornk) [inline tensorflow/core/kernels/crop_resize_bilinear_core.h]: After templating `load`, you can condense the comment. For example,
```
  // === loadx_ych ===
  // Loads top left and bottom left interpolation inputs into output argument
  // left. Loads top right and bottom right interpolation inputs into output
  // argument right. Pixels have y channels.
  // x can be 1, 2, 4, or 8:
  //   1: 4 left and right inputs are loaded with a single unaligned SSE load.
  //   2: 4 left inputs are loaded with first SSE load and 4 right inputs are 
  //      loaded with second SSE load.
  //   ...
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the fixes! I'm also wondering if we can use some code from Eigen. I'm through almost 2k lines of `crop_resize_bilinear_core.h` now (sorry again for the delay), and I have more high-level comments:
1. A lot of code are for packing and floating-point conversions: I wonder if you can reuse something from Eigen, like @rmlarsen said. Please see `<eigen_repo>/Eigen/src/Core/arch/{AVX,AVX512,SSE}`.
1. If we can't use primitives from Eigen, then these routines should be templated. 
    * `pack` can take vector type (`__m256i` vs `__m128i`), size of data type, and number of channels as template arguments. (And you can define a macro to select the intrinsics, e.g., `_MM_OR` as `_mm256_or_si256` for AVX2 or `_mm_or_si128` for AVX.) 
    * `load` and `extract_right` should also be templated similarly. 
    * We need to eliminate duplicate code as much as possible, otherwise it is going to be very hard to maintain/extend.
1. Most helper functions should be `inline`.
1. Could you please post the benchmark numbers? (Like https://github.com/tensorflow/tensorflow/pull/21681, for example.)

**Comment by penpornk** [comment]
@martinwicke Intrinsics are in `crop_resize_bilinear_core.h`. It is pretty big so github collapses the file by default.

**Comment by rmlarsen** [comment]
@martinwicke I exaggerated a bit, but it basically looks like a lot of plain C code.

File: tensorflow/core/kernels/crop_and_resize_op.cc (hunk 1 above)
Comment (johnpjf) [reply tensorflow/core/kernels/crop_and_resize_op.cc]: This one (crop and resize) wasn't actually updated yet. My plan was to make bilinear_resize accept a translate arg, then implement crop and resize with that.

**Comment by tfboyd** [comment]
NVIDIA will do an A/B test and determine if it is worth reworking this in Eigan or just closing.

**Comment by MattConley** [comment]
Here are the requested performance numbers run on my local machine with an Intel i7-5820K

Without the changes:
```
Benchmark                                           Time(ns) Iterations
-----------------------------------------------------------------------
BM_CropAndResize_cpu_DT_UINT8_1_640_640_3_512_512    6270880        100	 196.0M items/s
BM_CropAndResize_cpu_DT_UINT8_1_640_640_1_512_512    3954793        164	 103.6M items/s
BM_CropAndResize_cpu_DT_HALF_1_640_640_3_512_512     9292230        100	 132.2M items/s
BM_CropAndResize_cpu_DT_HALF_1_640_640_1_512_512     5378040        100	 76.2M items/s
BM_CropAndResize_cpu_DT_FLOAT_1_640_640_3_512_512    5738020        100	 214.2M items/s
BM_CropAndResize_cpu_DT_FLOAT_1_640_640_1_512_512    5039507        150	 81.3M items/s
BM_CropAndResize_cpu_DT_FLOAT_1_80_80_512_7_7         140624       4096	 932.1M items/s
```

With the changes:
```
Benchmark                                           Time(ns) Iterations
-----------------------------------------------------------------------
BM_CropAndResize_cpu_DT_UINT8_1_640_640_3_512_512    3932621        198  312.5M items/s 
BM_CropAndResize_cpu_DT_UINT8_1_640_640_1_512_512    1568988        427  261.1M items/s 
BM_CropAndResize_cpu_DT_HALF_1_640_640_3_512_512     6684210        100  183.8M items/s 
BM_CropAndResize_cpu_DT_HALF_1_640_640_1_512_512     2892926        204  141.6M items/s 
BM_CropAndResize_cpu_DT_FLOAT_1_640_640_3_512_512    2587333        267  474.9M items/s 
BM_CropAndResize_cpu_DT_FLOAT_1_640_640_1_512_512    1094434        585  374.3M items/s 
BM_CropAndResize_cpu_DT_FLOAT_1_80_80_512_7_7         132819       4054  741.5M items/s 
```

**Comment by penpornk** [comment]
@MattConley Thank you very much! :)

**Comment by tfboyd** [comment]
NVIDIA is closing based on not having internal experience with Eigen.  Would need to be picked up by TesnorFlow and maybe team up if needed.

**Pull request closed without merging**